src/content/en/fundamentals/getting-started/primers/async-functions.md

```diff
@@ -0,0 +1,345 @@
+      }
+
+      async getAvatar(name) {
+        const cache = await this.cachePromise;
+        return cache.match(`/avatars/${name}.jpg`);
+      }
+    }
+
+    const storage = new Storage();
+    storage.getAvatar('jaffathecake').then(…);
+
+Note: Class constructors cannot be async.
+
+## Example: Streaming a response
+
+The benefit of async functions increases in more complex examples. Say we wanted
+to stream a response while logging out the chunks and returning the final size.
+
+Note: The phrase "logging out the chunks" made me sick in my mouth.
+
+Here it is with promises:
+
+    function getResponseSize(url) {
+      return fetch(url).then(response => {
+        const reader = response.body.getReader();
+        let total = 0;
+
+        return reader.read().then(function processResult(result) {
+          if (result.done) return total;
+
+          const value = result.value;
+          total += value.length;
+          console.log('Received chunk', value);
+
+          return reader.read().then(processResult);
+        })
+      });
+    }
+
+Check me out, Jake "wielder of promises" Archibald. See how I'm calling
+`processResult` inside itself to set up an asynchronous loop? Writing that made
+me feel *very smart*. But like most "smart" code, you have to stare at it for
+ages to figure out what it's doing, like one of those magic-eye pictures from
+the 90's.
+
+Let's try that again with async functions:
+
+    async function getResponseSize(url) {
+      const response = await fetch(url);
+      const reader = response.body.getReader();
+      let result;
+      let total = 0;
+
+      while ((result = await reader.read()) && !result.done) {
+        const value = result.value;
+        total += value.length;
+        console.log('Received chunk', value);
+      }
+
+      return total;
+    }
+
+All the "smart" is gone. The asynchronous loop that made me feel so smug is
+replaced with a trusty, boring, while-loop. Much better. In future, we'll get
+[async iterators](https://github.com/tc39/proposal-async-iteration){:
+.external}, which would [replace the above `while`
+loop](https://gist.github.com/jakearchibald/0b37865637daf884943cf88c2cba1376){:
+.external}, making it even neater.
+
+Note: I'm sort-of in love with streams. If you're unfamiliar with streaming,
+[check out my
+guide](https://jakearchibald.com/2016/streams-ftw/#streams-the-fetch-api){:
+.external}.
+
+## Careful! Avoid going too sequential
+
+Although you're writing code that looks synchronous, ensure you don't miss the
+opportunity to do things in parallel.
+
+    async function series() {
+      await wait(500);
+      await wait(500);
+      return "done!";
+    }
+
+The above takes 1000ms to complete, whereas:
+
+    async function parallel() {
+      const wait1 = wait(500);
+      const wait2 = wait(500);
+      await wait1;
+      await wait2;
+      return "done!";
+    }
+
+…the above takes 500ms to complete, because both waits happen at the same time.
```

Comment:
We may want to call out `observables` are different than `promises` as they always start no matter what when you call them unlike `observables`.